### ISSUE TITLE
Bug: [no-unsafe-return] number returned in a callback of `sort` is reported as unsafe any

### ISSUE THREAD

**Issue by hasparus** [typescript-eslint/typescript-eslint#5364]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&jsx=true&sourceType=module&code=MYewdgzgLgBAHhGBeGBtA5AQ3QGhugI3QF0BuAKHIQDoIQAnKACiczwIEpkA+GAb3IwYoSLFABbAA6Z6AUwAmyGJmoAbEMEyrZAYRBSZsppwqCYcqAFd6YYfulz5pGAHoXMAKqRMAM1nnZKxsYEB9lWwADTDAATwiYKBjJBRgANy1LWWoYcgBfDnIgA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJFwEipCtTqMWDAPZU+DRLABmyKvGQ4+8BsMgB3BMZ4BfEFaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A

### Repro Code

```typescript
const xs = ['a', 'b'];

xs.sort((a, b) => {
  const compared = a.localeCompare(b);

  return compared; // Unsafe return of an `any` typed value. 
})

```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-return": "warn"
  }
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

As the tooltip on hover states that `compared` is a number, and `localeCompare` returns a number, I expect TypeScript ESLint rules to know it is in fact a number.

### Actual Result

`no-unsafe-return` reports that `compared` is any.

### Additional Info

- [Screenshot](https://user-images.githubusercontent.com/15332326/180505067-db50007e-7513-43a8-a62d-199da7ab49a3.png)

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.30.7` |
| `@typescript-eslint/parser` | `5.30.7` |
| `TypeScript` | `4.7.2` |
| `ESLint` | `8.15.0` |
| `node` | `web`, `16` |

---

_Edited to include [JoshuaKGoldberg](https://github.com/JoshuaKGoldberg)'s simplified repro._

**Comment by JoshuaKGoldberg** [comment]
Huh, very weird! This also happens if you simplify the code just a bit to always `return compared` [[simplified playground](pe=module&code=MYewdgzgLgBAHhGBeGBtA5AQ3QGhugI3QF0BuAKHIQDoIQAnKACiczwIEpkA+GAb3IwYoSLFABbAA6Z6AUwAmyGJmoAbEMEyrZAYRBSZsppwqCYcqAFd6YYfulz5pGAHoXMAKqRMAM1nnZKxsYEB9lWwADTDAATwiYKBjJBRgANy1LWWoYcgBfDnIgA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJFwEipCtTqMWDAPZU+DRLABmyKvGQ4+8BsMgB3BMZ4BfEFaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A]:

```ts
const xs = ['a', 'b'];

xs.sort((a, b) => {
  const compared = a.localeCompare(b);

  return compared; // Unsafe return of an `any` typed value. 
});
```

**Comment by hasparus** [comment]
Weird find, but assigning any of the arguments to an unused local variable cleared the error for me locally.

![image](https://user-images.githubusercontent.com/15332326/180509104-804755c8-776f-4b3c-97ef-e1f20fedc682.png)

I couldn't reproduce it in the playground, though 🤷‍♂️

**Comment by bradzacher** [comment]
This looks like a bug in TS?
Investigating in [ts-ast-viewer](https://ts-ast-viewer.com/#code/MYewdgzgLgBAHhGBeGBtAUDNByAhtgGhj3wF0DMcAjQ4qm89U9dBAOghACcoAKX3ESoBKZAD4YAb0qhIsUAFsADri4BTACbIYuVAAZSbADYhguI2oDCIZarW8q+0sJZYAlgDMYvRSvVaAQiQUPVFpLCx1KABXLjAYXzsNSgBfVxgo2PjdAEZDEzMLa1t1B1Q8lxThIA) - it looks like in TS 4.6 a regression was introduced which causes TS to report `compared` as having an `any` type.

However i can't reproduce this on the playground... very strange.

**Comment by JoshuaKGoldberg** [comment]
Ok, taking a deeper look, this looks to be a similar -or even the same- root issue as https://github.com/typescript-eslint/typescript-eslint/issues/4689 -> https://github.com/microsoft/TypeScript/issues/48313. Marking as external for now.

**Comment by awesomeunleashed** [comment]
I'm having an issue that's a bit more complex, but might be the same thing as this. Here's my reproduction repo: https://github.com/awesomeunleashed/ts-eslint-unsafe-bug-repro

You can see the error in the README there. In my actual code I started getting random `no-unsafe-xxx` warnings after some refactoring and they seemed to go away or come back based on seemingly unrelated changes in unrelated files.

If you guys could look if my case looks the same as the TS bug or like I should open a different issue when you have the time, that would be appreciated!

**Comment by mcmxcdev** [comment]
Also ran into this with similar (simplified) code:

```
{data.doctors
  .map((doctor) => doctor.npi)
  .join(', ')}
```

Hovering over in VSCode shows the correct types though.

**Comment by Josh-Cena** [comment]
On the latest TS version, this doesn't seem to be an error anymore.
